mantidproject/mantid
# Muon Load File Widget

closes #23322, 

This PR creates a new "load file widget" for use in the new Muon GUI. See the issue for a description of the main features.

Essentially the widget allows for a user to load a single or multiple files, via a Browse button or via explicitly entering the filenames into the widget.

As an extra feature, there is a copy-to-clipboard tool button for situations when the QLineEit becomes saturated with text, for example a list of 100 files.

Some files are duplicated from #23296 and do not need review here

- `threading_manager.py`
- `threading_manager_test.py`

Testing
=======

The unit tests on the presenter are a good reference point for the functionality I have explicitly coded into the widget.

The widget can be tested without Mantid using the following test code (for example add a file `test.py` into `/scripts/`)

```Python
from __future__ import (absolute_import, division, print_function)

from qtpy import QtGui, QtCore, QtWidgets
from mantid.simpleapi import *
from Muon.GUI.MuonAnalysis.loadfile.load_file_model_multithreading import BrowseFileWidgetModel
from Muon.GUI.MuonAnalysis.loadfile.load_file_view import BrowseFileWidgetView
from Muon.GUI.MuonAnalysis.loadfile.load_file_presenter import BrowseFileWidgetPresenter

from Muon.GUI.Common.muon_load_data import MuonLoadData

if __name__ == "__main__":
    import sys

    app = QtWidgets.QApplication(sys.argv)
    ui = BrowseFileWidgetPresenter(BrowseFileWidgetView(), BrowseFileWidgetModel(MuonLoadData()))
    ui.show()
    sys.exit(app.exec_())
```

Which will display the widget. Specific things to test:

- Try to cause a crash in any way possible.
- Try to load multiple files by using the browse button. Check the console output to ensure the Load algorithm is being run.
- Try to load multiple files by entering them by hand, as a ";" separated list.
- Check that the copy to clipboard icon works.
- Check that browsing for files with files already loaded **appends** the files to the list.
- Check that you can load more than 10 files, and that the behavior from this is understandable.

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
